cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -600,6 +600,7 @@ struct GrouperFastImpl : public Grouper {
   }
 
   Status Reset() override {
+    ARROW_DCHECK_EQ(temp_stack_.AllocatedSize(), 0);
```

Comment:
This is not related to the purpose of this PR but rather something remaining in https://github.com/apache/arrow/pull/41352#discussion_r1598969656 and too trivial for another PR.